src/Microsoft.AspNetCore.Mvc.Cors/Internal/CorsApplicationModelProvider.cs

```diff
@@ -67,24 +68,33 @@ public void OnProvidersExecuting(ApplicationModelProviderContext context)
 
                     if (isCorsEnabledGlobally || corsOnController || corsOnAction)
                     {
-                        UpdateHttpMethodActionConstraint(actionModel);
+                        UpdateActionToAcceptCorsPreflight(actionModel);
                     }
                 }
             }
         }
 
-        private static void UpdateHttpMethodActionConstraint(ActionModel actionModel)
+        private static void UpdateActionToAcceptCorsPreflight(ActionModel actionModel)
         {
             for (var i = 0; i < actionModel.Selectors.Count; i++)
             {
                 var selectorModel = actionModel.Selectors[i];
+
                 for (var j = 0; j < selectorModel.ActionConstraints.Count; j++)
                 {
                     if (selectorModel.ActionConstraints[j] is HttpMethodActionConstraint httpConstraint)
                     {
                         selectorModel.ActionConstraints[j] = new CorsHttpMethodActionConstraint(httpConstraint);
                     }
                 }
+
+                for (int j = 0; j < selectorModel.EndpointMetadata.Count; j++)
+                {
+                    if (selectorModel.EndpointMetadata[j] is IHttpMethodMetadata httpMethodMetadata)
```

Comment:
Hmm, should this be `is HttpMethodMetadata` instead of the interface to make sure we aren't replacing implementations we don't own?

Reply:
Yeah, I suggest matching on the exact type.